atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java

```diff
@@ -188,32 +190,67 @@ private SweepResults doRun(TableReference tableRef,
         }
     }
 
+    /**
+     * Returns batches with at least batchConfig.deleteBatchSize blocks per batch.
+     */
     private Iterator<BatchOfCellsToSweep> getBatchesToSweep(Iterator<List<CandidateCellForSweeping>> candidates,
                                                             SweepBatchConfig batchConfig,
                                                             SweepableCellFilter sweepableCellFilter,
                                                             ExaminedCellLimit limit) {
+        Iterator<List<CandidateCellForSweeping>> validCandidates =
+                Iterators.filter(candidates, list -> list != null && !list.isEmpty());
+
         Iterator<BatchOfCellsToSweep> cellsToSweep = Iterators.transform(
-                Iterators.filter(candidates, list -> !list.isEmpty()),
+                validCandidates,
                 sweepableCellFilter::getCellsToSweep);
         return new CellsToSweepPartitioningIterator(cellsToSweep, batchConfig.deleteBatchSize(), limit);
     }
 
     /**
-     * Returns the number of deleted (cell, timestamp) pairs.
+     * Partitions batches of blocks down to @param deleteBatchSize blocks per sweep deletion call.
+     * Returns the number of blocks that were deleted.
      */
-    private int sweepBatch(TableReference tableRef, List<CellToSweep> batch, RunType runType) {
-        Multimap<Cell, Long> startTimestampsToSweepPerCell = ArrayListMultimap.create();
-        List<Cell> sentinels = Lists.newArrayList();
+    private int sweepBatch(TableReference tableRef, List<CellToSweep> batch, RunType runType,
+            int deleteBatchSize) {
+        int numberOfSweptCells = 0;
+
+        Multimap<Cell, Long> currentBatch = ArrayListMultimap.create();
+        List<Cell> currentBatchSentinels = Lists.newArrayList();
+
         for (CellToSweep cell : batch) {
-            startTimestampsToSweepPerCell.putAll(cell.cell(), TDecorators.wrap(cell.sortedTimestamps()));
             if (cell.needsSentinel()) {
-                sentinels.add(cell.cell());
+                currentBatchSentinels.add(cell.cell());
+            }
+
+            if (currentBatch.size() + cell.sortedTimestamps().size() < deleteBatchSize) {
```

Comment:
I think we should use a larger number for the maximum batch size, e.g. `2 * deleteBatchSize` as we discussed before. The partitioning code will return batches that have at least `deleteBatchSize` (cell, ts) pairs, and we don't want each batch to be broken down into several delete() calls.